### ISSUE TITLE
Support disabling continue-on-failure mode

### ISSUE THREAD

**Issue by pekkaklarck** [robotframework/robotframework#4303]
When a keyword fails, Robot's default behavior is to stop executing the current test or a user keyword. There are, however, few cases where the remaining keywords in the test/keyword are run (i.e. the continue-on-failure mode is on):

1. When `robot:continue-on-failure` or `robot:recursive-continue-on-failure` tags have been used (#2285).
2. When executing keywords in test or keyword teardown.
3. When a test has a template, all top level iterations are run even if one of them would fail.

It would sometimes be convenient to be able to turn the continue-on-failure more off in the above cases. A concrete example is #4293, where TRY/EXCEPT behaves differently in teardowns and elsewhere. When `robot:recursive-continue-on-failure` was added, we [discussed](https://github.com/robotframework/robotframework/issues/2285#issuecomment-867528381) about adding `robot:no-continue-on-failure` to be able to disable it. It wasn't done then, but now would be a good time. For consistency reasons we should also support templates.

**Comment by oboehmer** [comment]
I can take this, @pekkaklarck ..

**Comment by pekkaklarck** [comment]
That would be great @oboehmer!

**Comment by pekkaklarck** [comment]
I think the only design decision related to this is deciding should `robot:no-continue-on-failure´ be recursive or not. I got a feeling it should be recursive even though it would be inconsistent with `robot:continue-on-failure`. Alternatively we could add separate `robot:recursive-no-continue-on-failure` but that's getting pretty complicated.

**Comment by oboehmer** [comment]
I would opt for it not to be recursive.. Not to continue is the default, so why should it be recursive? Let me create a PR shortly, we can discuss there..

**Comment by oboehmer** [comment]
please check #4306

**Comment by oboehmer** [comment]
one question is also if a test-level `robot:no-continue-on-failure` would propagate into the Teardown method or not?

```
No-continue-on-failure-2 in Teardown
    [Documentation]    FAIL 1
    [Tags]   robot:no-continue-on-failure
    [Teardown]   Run Keywords   Fail   1   AND   Fail  2
    Log    does not matter
```

**Comment by pekkaklarck** [comment]
The reason I got a feeling `robot:no-continue-on-failure` should be recursive is that then you could just set it once for the teardown keyword. Otherwise you needed to set it to all keywords used in teardowns to make sure TRY/EXCEPT works consistently (#4293). Same thing if you want to stop `robot:recursive-continue-on-failure` from propagating. It would be odd if continue-on-failure mode would be off in a keyword using `robot:no-continue-on-failure` but would again be enabled in keywords used by it.

`robot:no-continue-on-failure` being recursive when ´robot:continue-on-failure` is not is obviously somewhat inconsistent. Possible solutions:
- Add both `robot:no-continue-on-failure` and  `robot:recursive-no-continue-on-failure` to match the earlier added `robot:continue-on-failure` and `robot:recursive-continue-on-failure`. That `robot:recursive-no-continue-on-failure´ is a rather long tag name but this would be a pretty explicit, consistent and backwards incompatible solution.
- Change `robot:continue-on-failure` to be recursive and deprecate `robot:recursive-continue-on-failure`. Having just `robot:continue-on-failure` and `robot:no-continue-on-failure` would be simpler but not being able to limit continue-on-failure only to a certain level could be problematic. This would obviously be a backwards incompatible change as well.

**Comment by pekkaklarck** [comment]
I think test level `robot:no-continue-on-failure` _should not_ affect teardowns. We can think teardowns having implicit `robot:recursive-continue-on-failure` set and it overrides possible higher level configuration.

**Comment by pekkaklarck** [comment]
Sorry, don't have time to look at PRs today. It would be important to agree on the semantics first anyway. That might be something to discuss further on Slack.

**Comment by oboehmer** [comment]
agree, started a thread on slack..

**Comment by pekkaklarck** [comment]
I've been thinking this a bit more and here's how I currently think we should do this:

- Let's use `robot:stop-on-failure` instead of `robot:no-continue-on-failure`.
- Make it not recursive to match how `robot:continue-on-failure` works.
- Add also `robot:recursive-stop-on-failure` to match `robot:recursive-continue-on-failure`.
- Neither `stop-on-failure` variants should propagate to teardowns. We can think them having `robot:recursive-continue-on-failure` implicitly. To enable the stop-on-failure mode in teardowns, these special tags need to be used explicitly.

**Comment by oboehmer** [comment]
@pekkaklarck , I created PR #4313 based on the recent design discussions. Please review when you have a chance.

**Comment by pekkaklarck** [comment]
Thanks for the PR! I won't have time to look at it in the immediate future but hopefully next week. I want to do few things in master before we create a maintenance branch for RF 5.0.x and start merging RF 5.1 features.

**Comment by pekkaklarck** [comment]
This was implemented in PR #4313 by @oboehmer. Thanks a great contribution!
